packages/api/lambdas/ems-distribution-report.js

```diff
@@ -0,0 +1,179 @@
+      this.rawLine.split('[')[1].split(']')[0],
+      'DD/MMM/YYYY:HH:mm:ss ZZ'
+    );
+  }
+
+  get transferStatus() {
+    return this.bytesSent === this.objectSize ? 'S' : 'F';
+  }
+
+  get username() {
+    const requestUri = this.rawLine.split('"')[1].split(' ')[1];
+    const parsedUri = (new URL(requestUri, 'http://localhost'));
+    return parsedUri.searchParams.get('x-EarthdataLoginUsername');
+  }
+
+  /**
+   * Return the event in an EMS-parsable format
+   *
+   * @returns {string} an EMS distribution log entry
+   */
+  toString() {
+    return [
+      this.time.toISOString(),
+      this.username,
+      this.remoteIP,
+      `s3://${this.bucket}/${this.key}`,
+      this.bytesSent,
+      this.transferStatus
+    ].join('|&|');
+  }
+}
+
+async function fetchEventsFromObject(getObjectParams) {
+  log.info(`Fetching events from s3://${getObjectParams.Bucket}/${getObjectParams.Key}`);
+
+  // Fetch the S3 Server Log object from S3
+  const getObjectResponse = await aws.s3().getObject(getObjectParams).promise();
+
+  return getObjectResponse.Body
+    // Get the contents of the S3 object
+    .toString()
+    // Break the file into separate lines
+    .split('\n')
+    // Remove lines that aren't for distribution events
+    .filter(DistributionEvent.isDistributionEvent)
+    // Convert the remaining lines into DistributionEvent objects
+    .map((logLine) => new DistributionEvent(logLine));
+}
+
+async function fetchEvents(sourceBucket, startTime, endTime) {
+  log.info(`Fetching distribution logs from ${startTime.toISOString()} to ${endTime.toISOString()}`); // eslint-disable-line max-len
+
+  // Fetch all of the log objects in the S3 bucket
+  const s3Objects = await aws.listS3ObjectsV2({ Bucket: sourceBucket });
+  // We only want the Bucket and Key properties
+  const simpleS3Objects = s3Objects.map((s3Object) =>
+    ({ Bucket: sourceBucket, Key: s3Object.Key }));
+
+  // Throttle how many log objects we fetch and parse in parallel
+  const queue = new TaskQueue(Promise, 5);
+  const throttledFetchEventsFromObject = queue.wrap(fetchEventsFromObject);
+
+  // A few utility functions that we'll be using below
+  const flatten = (accumulator, currentValue) => accumulator.concat(currentValue);
+  const timeFilter = (event) => event.time >= startTime && event.time < endTime;
+  const sortByTime = (eventA, eventB) => (eventA.time < eventB.time ? -1 : 1);
+
+  // Fetch all distribution events from S3
+  return (await Promise.all(simpleS3Objects.map(throttledFetchEventsFromObject)))
+    // Flatten the result arrays into a single array
+    .reduce(flatten, [])
+    // Remove results outside of the specified timerange
+    .filter(timeFilter)
+    // Sort the results
+    .sort(sortByTime);
+}
+
+function storeDistributionReport(params) {
+  const {
+    destinationBucket,
+    distributionReport,
+    startTime,
+    endTime
+  } = params;
+
+  const destinationKey = `${startTime.toISOString()}_to_${endTime.toISOString()}.log`;
```

Comment:
The filename should be `<YYYYMMDD> _<Provider>_<FileType>_<DataSource>.flt.rev<1- n>`
It's in the page 46 of EMS document.

Reply:
If the report is generated more than once, .rev extension should be added, since EMS doesn't accept the same file again.